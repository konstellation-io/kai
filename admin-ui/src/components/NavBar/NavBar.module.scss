@import '../../styles/colors';
@import '../../styles/mixins';
@import '../../styles/variables';

.container {
  border: solid 0px black;
  border-width: 1px 0;
}
.container a:not(:last-child) div {
  border: solid 0px black;
  border-bottom-width: 1px;
}

.disabled {
  cursor: default;
}

.item {
  @include font-button;
  text-align: center;
  color: $font-color;
  line-height: 10px;

  height: 56px;
  position: relative;
  padding: 0 28px;
  white-space: nowrap;
  cursor: pointer;
  transition: color ease 0.4s, background-color ease 0.4s;

  display: flex;
  flex-direction: row;
  align-items: center;

  span {
    margin: 0 15px;
  }

  .warning {
    color: $color-orange;
  }

  &:hover {
    background-color: rgba(191, 207, 255, 0.04);
    color: $font-color-white;
  }

  &.disabled {
    pointer-events: none;
    color: $font-color-dark;
  }
}

.active .item {
  background-color: rgba(191, 207, 255, 0.15);
  color: white;
}
