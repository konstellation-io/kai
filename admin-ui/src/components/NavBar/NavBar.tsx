import { has } from 'lodash';

import React from 'react';
import IconWarning from '@material-ui/icons/Warning';
import { NavLink } from 'react-router-dom';

import styles from './NavBar.module.scss';
import cx from 'classnames';

export type Tab = {
  label: string;
  route: string;
  Icon?: any;
  exact?: boolean;
  disabled?: boolean;
  showWarning?: boolean;
  warningTitle?: string;
};
type Props = {
  tabs: Tab[];
  defaultActive?: number;
};

function NavBar({ tabs }: Props) {
  function handleClick(event: any, disabled: boolean = false) {
    if (disabled) {
      event.preventDefault();
    }
  }

  const tabElements = tabs.map((tab, idx) => {
    const exact = has(tab, 'exact') ? tab.exact : true;
    return (
      <NavLink
        key={`NavBarItem_${idx}`}
        to={tab.route}
        activeClassName={styles.active}
        exact={exact}
        onClick={e => handleClick(e, tab.disabled)}
        className={cx({ [styles.disabled]: tab.disabled })}
        replace
      >
        <div className={cx(styles.item, { [styles.disabled]: tab.disabled })}>
          <tab.Icon className="icon-regular" />
          <span>{tab.label}</span>
          {tab.showWarning && (
            <div title={tab.warningTitle}>
              <IconWarning className={cx('icon-regular', styles.warning)} />
            </div>
          )}
        </div>
      </NavLink>
    );
  });

  return <div>{tabElements}</div>;
}

export default NavBar;
