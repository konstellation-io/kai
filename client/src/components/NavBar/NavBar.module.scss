.container {
  border: solid 0px black;
  border-width: 1px 0;
}
.container a:not(:last-child) div {
  border: solid 0px black;
  border-bottom-width: 1px;
}
