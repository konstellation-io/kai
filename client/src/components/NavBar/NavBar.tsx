import React, {useState, useEffect} from 'react';

import Button, { BUTTON_TYPES, BUTTON_ALIGN } from '../Button/Button';
import { IconProp } from '@fortawesome/fontawesome-svg-core';

import styles from './NavBar.module.scss';


export type Tab = {
  label: string,
  icon?: IconProp,
  route?: string,
}
type Props = {
  tabs: Tab[];
  defaultActive?: number;
  onChange?: Function;
};

function NavBar({
  tabs,
  defaultActive = 0,
  onChange = function(idx:number) {}
}: Props) {
  const [activeTab, setActiveTab] = useState(defaultActive);

  useEffect(() => {
    setActiveTab(defaultActive);
  }, [defaultActive]);

  const tabElements = tabs.map((tab, idx) => (
    <Button
      key={`tabButton${idx}`}
      label={tab.label}
      icon={tab.icon}
      primary={activeTab === idx}
      type={ BUTTON_TYPES.TRANSPARENT }
      height={56}
      onClick={() => {
        setActiveTab(idx);
        onChange(idx);
      }}
      align={BUTTON_ALIGN.LEFT}
      to={tab.route}
    />
  ));

  return (
    <div className={styles.container}>
      {tabElements}
    </div>
  );
}

NavBar.propTypes = {
};

export default NavBar;
