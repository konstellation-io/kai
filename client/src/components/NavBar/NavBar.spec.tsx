import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import NavBar, {Tab} from './NavBar';
import '@testing-library/jest-dom/extend-expect';

const tabs = [
  {label: 'TAB 1'} as Tab,
  {label: 'TAB 2'} as Tab,
  {label: 'TAB 3'} as Tab
];

afterEach(cleanup);

it('Renders NavBar without crashing', () => {
  const { container } = render(<NavBar />);
  expect(container).toMatchSnapshot();
});

it('Shows different options', () => {
  const { getByText } = render(<NavBar tabs={tabs} />);

  expect(getByText('TAB 1')).toBeInTheDocument();
  expect(getByText('TAB 2')).toBeInTheDocument();
  expect(getByText('TAB 3')).toBeInTheDocument();
});

it('Highlights selected option', () => {
  const { container } = render(<NavBar tabs={tabs} defaultActive={1} />);

  const primary = container.querySelector('.primary');
  expect(primary && primary.textContent).toBe('TAB 2');
});

it('Handles click events', () => {
  const clickMock = jest.fn(() => true);

  const { container, getByText } = render(
    <NavBar tabs={tabs} defaultActive={0} onChange={clickMock}/>
  );

  let primary = container.querySelector('.primary');
  expect(primary && primary.textContent).toBe('TAB 1');

  fireEvent.click(getByText('TAB 3'));

  primary = container.querySelector('.primary');
  expect(primary && primary.textContent).toBe('TAB 3');
  expect(clickMock).toHaveBeenCalledTimes(1);
});

it('Updates active tab', () => {
  const { container, rerender } = render(
    <NavBar tabs={tabs} defaultActive={0} />
  );

  let primary = container.querySelector('.primary');
  expect(primary && primary.textContent).toBe('TAB 1');

  rerender(<NavBar tabs={tabs} defaultActive={1} />);

  primary = container.querySelector('.primary');
  expect(primary && primary.textContent).toBe('TAB 2');
});
