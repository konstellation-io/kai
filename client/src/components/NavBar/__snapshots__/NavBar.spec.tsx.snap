// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders NavBar without crashing 1`] = `
<div>
  <div
    class="container"
  />
</div>
`;
